src/computability/encoding.lean

```diff
@@ -0,0 +1,189 @@
+/-
+Copyright (c) 2020 Pim Spelier, Daan van Gent. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Pim Spelier, Daan van Gent.
+-/
+
+import data.fintype.basic
+import data.num.lemmas
+import tactic
+
+/-!
+# Encodings
+
+This file contains the definition of a (finite) encoding, a map from a type to
+strings in an alphabet, used in defining computability by Turing machines.
+It also contains several examples:
+
+## Examples
+
+- `fin_encoding_nat_Γ₀₁`   : a binary encoding of ℕ in a simple alphabet.
+- `fin_encoding_nat_Γ'`    : a binary encoding of ℕ in the alphabet used for TM's.
+- `unary_fin_encoding_nat` : a unary encoding of ℕ
+- `fin_encoding_bool_Γ₀₁`  : an encoding of bool.
+-/
+
+namespace computability
+
+/-- An encoding of a type in a certain alphabet, together with a decoding. -/
+structure encoding (α : Type) :=
+(Γ : Type)
+(encode : α → list Γ)
+(decode : list Γ → option α)
+(encodek : ∀ x, decode (encode x) = some x)
```

Comment:
Why the name `encodek`? Perhaps `decode_encode` would be closer to the recommended naming scheme.

Reply:
We went for the same name for this property as in the definition of `encodable`, the similar construct that was already in Lean but isn't suited to explicit work with Turing machines. I agree, `decode_encode` does seem to be closer to the usual naming scheme. (Or maybe even `decode_encode_eq_some`?)

Reply:
`encodek` is inherited from the stone ages, I have no idea what it abbreviates. I think `decode_encode` is a better name, although I appreciate the brevity of `encodek` because it is frequently referenced.